controllers/wordpressinstance_controller.go

```diff
@@ -50,283 +59,90 @@ type WordpressInstanceReconciler struct {
-	//     same label on the instance. For later querying
-	// Maybe put the IP of the load balancer in the wordpress instance CRD
-	// Represent the status of the instance in the CRD?
-	// Use defaults
-	r.Log.V(2).Info("BEFORE extract objects")
-	objects, err := r.ExtractObjects(ctx, &tmplOutput)
+	endpoint, err := GetEndpoint(ctx, r.Client, *wp)
 	if err != nil {
-		r.Log.V(0).Info("Error extracting objects!", "err", err)
+		return ctrl.Result{RequeueAfter: shortWait}, err
 	}
-
-	err = r.createObjects(ctx, objects, i)
-
-	if err != nil {
-		r.Log.V(0).Info("Error creating objects!", "err", err)
+	if endpoint != "" {
+		wp.Status.Endpoint = endpoint
 	}
-
-	return ctrl.Result{}, nil
-}
-
-func (r *WordpressInstanceReconciler) ExtractObjects(ctx context.Context, s *string) ([]*unstructured.Unstructured, error) {
-	// read full output from job by retrieving the logs for the job's pod
-	r.Log.V(2).Info("ENTER extract objects", "ctx", ctx, "string", s)
-	reader := strings.NewReader(*s)
-
-	// decode and process all resources from job output
-	d := yaml.NewYAMLOrJSONDecoder(reader, 4096)
-	var objects []*unstructured.Unstructured
-	for {
-		obj := &unstructured.Unstructured{}
-		if err := d.Decode(&obj); err != nil {
-			if err == io.EOF {
-				// we reached the end of the job output
-				r.Log.V(2).Info("EXIT extract objects because EOF", "objects", objects, "err", err)
-				break
-			}
-			r.Log.V(2).Info("EXIT extract objects because ERROR", "objects", objects, "err", err)
-			return nil, errors.Wrapf(err, "failed to parse output")
-		}
-
-		objects = append(objects, obj)
+	// Update is no-op if there is no change on the object.
+	if err := r.Client.Update(ctx, wp); err != nil {
+		return ctrl.Result{RequeueAfter: shortWait}, err
 	}
-
-	r.Log.V(2).Info("EXIT extract objects", "objects", objects)
-	return objects, nil
+	return ctrl.Result{RequeueAfter: longWait}, nil
 }
 
-func (r *WordpressInstanceReconciler) createObjects(ctx context.Context, objects []*unstructured.Unstructured, i *wordpressv1alpha1.WordpressInstance) error {
-	for _, obj := range objects {
-		// process and create the object that we just decoded
-		if err := r.createOutputObject(ctx, obj, i); err != nil {
-			return err
-		}
+func ApplyResources(ctx context.Context, kube client.Client, wp wordpressv1alpha1.WordpressInstance) error {
+	database := GetMySQLInstance(wp)
+	if err := kube.Patch(ctx, &database, client.Apply); err != nil && !kerrors.IsAlreadyExists(err) {
+		return errors.Wrap(err, fmt.Sprintf("cannot create %s/%s in namespace %s", database.Kind, database.GetName(), database.GetNamespace()))
+	}
+	cluster := GetKubernetesCluster(wp)
+	if err := kube.Patch(ctx, &cluster, client.Apply); err != nil && !kerrors.IsAlreadyExists(err) {
+		return errors.Wrap(err, fmt.Sprintf("cannot create %s/%s in namespace %s", cluster.Kind, cluster.GetName(), cluster.GetNamespace()))
+	}
+	kubernetesApplication := GetKubernetesApplication(wp, database.GetName())
```

Comment:
I like the general direction we're going of making things nicer and easier to work with. I'm trying to think about how we can also make the code as easy to follow as possible; yaml may not be very nice to work with, but it's pretty easy to follow the flow of (since there is no flow, haha).

Anyway, one adjustment which I think could make sense is to the way that the names are passed around, so that the data only flows in one direction. I would recommend that we create the names *outside* of resources.go, at the top of this function. Then we can inject them into all of the templating-equivalent methods. Specifically, I feel that the way we are currently creating the database name in `GetMySQLInstance`, then returning an object, then threading that object's name into `GetKubernetesApplication`, is a bit hard to follow at first. But it could just be me. What do you think? My real goal is to make reading the code as simple as possible, so it could be the case that doing this would also work just as well:

```go
database := GetMySQLInstance(wp)
cluster := GetKubernetesCluster(wp)
kubernetesApplication := GetKubernetesApplication(wp, database.GetName())

// then, call patch a bunch of times and collect the errors
// then, return
```